textx/scoping/rrel.py

```diff
@@ -0,0 +1,475 @@
+        return self.path_elements[0].start_locally()
+
+    def start_at_root(self):
+        return self.path_elements[0].start_at_root()
+
+
+class RRELExpression:
+    def __init__(self, seq, importURI):
+        self.seq = seq
+        self.importURI = importURI
+
+    def __repr__(self):
+        return "+m:" + str(self.seq) if self.importURI else str(self.seq)
+
+
+class RRELVisitor(PTNodeVisitor):
+
+    def visit_rrel_parent(self, node, children):
+        return RRELParent(children[0])
+
+    def visit_rrel_navigation(self, node, children):
+        if len(children) == 1:
+            return RRELNavigation(children[0], True)
+        else:
+            return RRELNavigation(children[1], False)
+
+    def visit_rrel_brackets(self, node, children):
+        assert(len(children) == 1)  # a path
+        return RRELBrackets(children[0])
+
+    def visit_rrel_dots(self, node, children):
+        return RRELDots(len(node.value))
+
+    def visit_rrel_zero_or_more(self, node, children):
+        return RRELZeroOrMore(children[0])
+
+    def visit_rrel_path(self, node, children):
+        return RRELPath(children)
+
+    def visit_rrel_sequence(self, node, children):
+        return RRELSequence(children)
+
+    def visit_rrel_path_element(self, node, children):
+        assert(len(children) == 1)
+        return children[0]
+
+    def visit_rrel_expression(self, node, children):
+        if len(children) == 1:
+            return RRELExpression(children[0], False)
+        else:
+            return RRELExpression(children[1], True)
+
+
+def parse(rrel_expression):
+    """
+    This function parses a rrel path and returns a RREL expression tree.
+
+    Args:
+        rrel_expression: the RREL expression (string).
+
+    Returns:
+        A RREL expression tree.
+    """
+    from arpeggio import ParserPython
+    parser = ParserPython(rrel_standalone, reduce_tree=False)
+    parse_tree = parser.parse(rrel_expression)
+    return visit_parse_tree(parse_tree, RRELVisitor())
+
+
+def find(obj, lookup_list, rrel_tree, obj_cls=None):
+    """
+    This function gets all/one element from a model
+    object based on an rrel tree (query).
+
+    Note: this function uses Python 3.3+ features (recursive generators)
+
+    Args:
+        obj: model object (starting point of the search)
+        lookup_list: list of name parts forming the full name.
+        rrel_tree: the query (must be a RRELExpression object or a string)
+
+    Returns:
+        The result of the query (first match), a
+        Postponed object, or None (nothing found)
+    """
+    from textx import textx_isinstance
+    from textx.scoping import Postponed
+    if isinstance(rrel_tree, string_types):
+        rrel_tree = parse(rrel_tree)
+    assert isinstance(rrel_tree, RRELExpression)
+    rrel_tree = rrel_tree.seq
+    assert isinstance(rrel_tree, RRELSequence)
+    if isinstance(lookup_list, string_types):
+        lookup_list = lookup_list.split(".")
```

Comment:
We should probably find another way to split the lookup name as this assume that `.` will always be used which is generally not true.

A quick thought. A grammar rule might provide this info. Something like:
```
FQN[split='.']: ID+['.'];
```

Reply:
Very good idea!

Reply:
A first experiment (to be discussed):
  * Allow the `split` parameter
  * Change `create_rrel_scope_provider` to accept a split_string directly (e.g. '.' or '::') or a rule_name used to extract the split_string from the meta model, once all rules are parsed.
  * Unittests adapted (https://github.com/textX/textX/pull/274/commits/14df5467fcf4ed731780c475bbdc55c138add323):
     * `test_referencing_attributes_with_rrel_all_in_one_splitstring`: grammar/rule based spli string extraction
     * `test_rrel.py`: find-API test with fixed split string.

Open: 
  * The grammar/RREL expression could contain an option to specify the split_string, like `+m:` for importURI, `+split('::'):`.

Reply:
@goto40 Excellent! I have another proposal: since we have now `+m` as a expression marker which changes the lookup algorithm, we should probably to the same with `^` as they belong to the same kind of operators. Maybe instead of `^` to use something like `+u` for upward search so we could say: `+m+u:...` instead of `+m:^...`. I think this will be more consistent and will open a possibility to introduce more expression markers in the future. So all expression markers will be of the same form `+<letter>` and will come before `:`. 

One more thing to discuss is `~` and `*`. One is prefix and the other is in postfix form. Should we leave it that way? Would it be more consistent to use postfix form for both (or prefix)? Instead of `~extends*` to write `extends~*` (or `extends*~`)? What is more logical, easier to comprehend?

Reply:
In general, I agree. However, for '^' we have a different case: you can have it for one path in a sequence and not for another. So it is not really global. Note: '^` is only a shortcut for' (..)*'. 

For '~' : if we use a postfix '(... )*~' is difficult to interpret...

Reply:
Of we could later add options like ignore-case, etc.

Reply:
Ok, sounds good. Agree on both points.